### PR TITLE
feat(query):support array_indexof

### LINKED ISSUE #9839: array_indexof(arr, item)

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

indexof(list, element) | Returns the index of the element if the list contains the element. | indexof([1, 2, NULL], 2) | 2
-- | -- | -- | --

Closes #9839

### CODE REVIEW

File: src/query/functions/src/scalars/array.rs [hunk 1]
```diff
@@ -149,6 +149,30 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+        "indexof",
+        FunctionProperty::default(),
+        |_, _| FunctionDomain::Full,
+        vectorize_with_builder_2_arg::<ArrayType<NullableType<GenericType<0>>>, GenericType<0>, UInt64Type>(
+            |arr, val, output, _| {
+                let len = arr.len();
+                let mut matched = false;
+                for idx in 0..len {
+                    if let Some(Some(item)) = arr.index(idx) {
+                        if item == val {
```
Comment (b41sh) [inline src/query/functions/src/scalars/array.rs:162]: How should we handle the case where val is `NULL`?

Comment (TCeason) [reply src/query/functions/src/scalars/array.rs:162]: Directly return 0

File: src/query/functions/src/scalars/array.rs [hunk 2]
```diff
@@ -149,6 +149,30 @@ pub fn register(registry: &mut FunctionRegistry) {
         ),
     );
 
+    registry.register_2_arg_core::<ArrayType<NullableType<GenericType<0>>>, GenericType<0>, UInt64Type, _, _>(
```
[the remark this answers is not in the document]
Comment (sundy-li) [reply src/query/functions/src/scalars/array.rs:152]: Yes, array_concat had this redundant register

File: src/query/functions/src/scalars/array.rs (hunk 1 above)
Comment (b41sh) [reply src/query/functions/src/scalars/array.rs:162]: 0 is not a valid index, how about return `NULL`?

Comment (TCeason) [reply src/query/functions/src/scalars/array.rs:162]: Now is this:
```
select indexof([1,2,null,2,null], null)
----
3
```
I think that's better. :+1:

File: src/query/functions/src/scalars/array.rs [hunk 3]
```diff
@@ -149,6 +149,30 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+                        if item == val {
+                            output.push(idx as u64 + 1);
+                            matched = true;
+                            break;
+                        }
+                    }
+                }
+                if !matched {
+                    output.push(0u64)
+                }
```
Comment (andylokandy) [inline src/query/functions/src/scalars/array.rs:171]:
```suggestion
                            return;
                        }
                    }
                }
                
                output.push(0u64);
```

File: src/query/functions/src/scalars/array.rs [hunk 4]
```diff
@@ -149,6 +149,30 @@ pub fn register(registry: &mut FunctionRegistry) {
         ),
     );
 
+    registry.register_2_arg_core::<ArrayType<GenericType<0>>, GenericType<0>, UInt64Type, _, _>(
```
Comment (andylokandy) [inline src/query/functions/src/scalars/array.rs:152]: Why not use `register_passthrough_nullable_2_arg`?

Comment (andylokandy) [inline src/query/functions/src/scalars/array.rs:152]: May need to add a implementaion for `EmptyArray`

File: src/query/functions/tests/it/scalars/array.rs [hunk 5]
```diff
@@ -150,3 +153,29 @@ fn test_concat(file: &mut impl Write) {
     );
     run_ast(file, "concat([1,2,null], [int8_col])", &columns);
 }
+
+fn test_indexof(file: &mut impl Write) {
+    run_ast(file, "indexof([false, true], false)", &[]);
+    run_ast(file, "indexof([], false)", &[]);
```
Comment (andylokandy) [inline src/query/functions/tests/it/scalars/array.rs:159]: add a case for `run_ast(file, "indexof([], NULL)", &[]);`

File: src/query/functions/tests/it/scalars/array.rs [hunk 6]
```diff
@@ -130,6 +131,8 @@ fn test_contains(file: &mut impl Write) {
 }
 
 fn test_concat(file: &mut impl Write) {
+    run_ast(file, "concat([], [1,2])", &[]);
```
Comment (andylokandy) [inline src/query/functions/tests/it/scalars/array.rs:134]: suggestion: add a case for `concat([], [])`

**Review by b41sh — APPROVED** [review]

**Pull request merged**